src/os/bluestore/Writer.cc

```diff
@@ -0,0 +1,1384 @@
+  bool left_do_pad;
+  bool right_do_pad;
+  uint32_t left_location;
+  uint32_t right_location;
+  if (p2phase(location, au_size) != 0) {
+    blob_data_t& first_blob = blobs.front();
+    if (!first_blob.is_compressed()) {
+      // try to make at least disk block aligned
+      std::tie(left_do_pad, left_location) = _write_expand_l(location);
+      if (left_location < location) {
+        bufferlist tmp;
+        if (left_do_pad) {
+          tmp.append_zero(location - left_location);
+        } else {
+          tmp = _read_self(left_location, location - left_location);
+        }
+        tmp.claim_append(first_blob.disk_data);
+        first_blob.disk_data.swap(tmp);
+        first_blob.real_length += location - left_location;
+        location = left_location;
+      }
+    }
+  }
+  if (p2phase(data_end, au_size) != 0) {
+    blob_data_t& last_blob = blobs.back();
+    if (!last_blob.is_compressed()) {
+    // try to make at least disk block aligned
+      std::tie(right_do_pad, right_location) = _write_expand_r(data_end);
+      if (data_end < right_location) {
+        // TODO - when we right-expand because of some blob csum restriction, it is possible
+        // we will be left-blob-csum-unaligned. It is wasted space.
+        // Think if we want to fix it.
+        if (right_do_pad) {
+          last_blob.disk_data.append_zero(right_location - data_end);
+        } else {
+          bufferlist tmp;
+          tmp = _read_self(data_end, right_location - data_end);
+          last_blob.disk_data.append(tmp);
+        }
+        last_blob.real_length += right_location - data_end;
+      }
+      data_end = right_location;
+    }
+  }
+}
+
+// Writes uncompressed data.
+void BlueStore::Writer::do_write(
+  uint32_t location,
+  bufferlist& data)
+{
+  do_deferred = false;
+  disk_allocs.it = allocated.end();
+  disk_allocs.pos = 0;
+  dout(20) << __func__ << " 0x" << std::hex << location << "~" << data.length() << dendl;
+  dout(25) << "on: " << onode->print(pp_mode) << dendl;
+  blob_vec bd;
+  uint32_t ref_end = location + data.length();
+  uint32_t data_end = location + data.length();
+  _split_data(location, data, bd);
+  _align_to_disk_block(location, data_end, bd);
+  if (ref_end < onode->onode.size) {
+    ref_end = std::min<uint32_t>(data_end, onode->onode.size);
+  }
+  dout(20) << "blobs to put:" << blob_data_printer(bd, location) << dendl;
+  statfs_delta.stored() += ref_end - location;
+  exmp_it after_punch_it =
+    bstore->_punch_hole_2(onode->c, onode, location, data_end - location,
+    released, pruned_blobs, shared_changed, statfs_delta);
+  dout(25) << "after punch_hole_2: " << std::endl << onode->print(pp_mode) << dendl;
+
+  uint32_t au_size = bstore->min_alloc_size;
+  if (au_size != bstore->block_size) {
+    _try_put_data_on_allocated(location, data_end, ref_end, bd, after_punch_it);
+  }
+  if (location != data_end) {
+    uint32_t need_size = p2roundup(data_end, au_size) - p2align(location, au_size);
```

Comment:
dumb question: what if need_size is too big it conflicts with other writes causing a enospc?

Reply:
This particular code got changed under pressure of compressed blobs.
Now `need_size` is calculated differently.